sklearn/multioutput.py

```diff
@@ -359,3 +362,224 @@ def score(self, X, y):
+        list of indices in the order attribute.
+
+    References
+    ----------
+    Jesse Read, Bernhard Pfahringer, Geoff Holmes, Eibe Frank, "Classifier
+    Chains for Multi-label Classification", 2009.
+
+    """
+    def __init__(self, base_estimator, order=None, cv=None, random_state=None):
+        self.base_estimator = base_estimator
+        self.order = order
+        self.cv = cv
+        self.random_state = random_state
+
+    def fit(self, X, Y):
+        """Fit the model to data matrix X and targets Y.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+        Y : {array-like, sparse matrix}, shape (n_samples, n_classes)
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        X, Y = check_X_y(X, Y,
+                         multi_output=True,
+                         accept_sparse=True)
+
+        random_state = check_random_state(self.random_state)
+
+        self.order_ = self.order
+        if self.order_ is None:
+            self.order_ = np.array(range(Y.shape[1]))
+        elif isinstance(self.order_, str):
+            if self.order_ == 'random':
+                self.order_ = random_state.permutation(Y.shape[1])
+        elif sorted(self.order_) != list(range(Y.shape[1])):
+                raise ValueError("invalid order")
+
+        self.estimators_ = [clone(self.base_estimator)
+                            for _ in range(Y.shape[1])]
+
+        self.classes_ = []
+
+        if self.cv is None:
+            Y_pred_chain = Y[:, self.order_]
+        else:
+            Y_pred_chain = np.zeros((X.shape[0], len(self.estimators_)))
+
+        if sp.issparse(X):
+            X_aug = sp.hstack((X, Y_pred_chain), format='lil')
+        else:
+            X_aug = np.hstack((X, Y_pred_chain))
+
+        for chain_idx, estimator in enumerate(self.estimators_):
+            y = Y[:, self.order_[chain_idx]]
+            estimator.fit(X_aug[:, :(X.shape[1] + chain_idx)], y)
+            if self.cv is not None and chain_idx < len(self.estimators_) - 1:
+                col_idx = X.shape[1] + chain_idx
+                cv_result = cross_val_predict(
+                    self.base_estimator, X_aug[:, :col_idx],
+                    y=y, cv=self.cv)
+                if sp.issparse(X_aug):
+                    X_aug[:, col_idx] = np.expand_dims(cv_result, 1)
+                else:
+                    X_aug[:, col_idx] = cv_result
+
+            self.classes_.append(estimator.classes_)
+        return self
+
+    def predict(self, X):
+        """Predict on the data matrix X using the ClassifierChain model.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+
+        Returns
+        -------
+        Y_pred : array-like, shape (n_samples, n_classes)
+        """
+        Y_pred_chain = np.zeros((X.shape[0], len(self.estimators_)))
```

Comment:
Need to do check_array before X.shape. We need to test that this works with lists of lists. For accept_sparse, CSC may be appropriate.